base/applications/cleanmgr/util.c

```diff
@@ -0,0 +1,1120 @@
+{
+    WCHAR *StrNum = NULL;
+    WCHAR *ValStr = NULL;
+    
+    ValStr = (PWCHAR)malloc(ARR_MAX_SIZE);
+    
+    int Num = 0;
+    
+    if (nArgs == 2)
+    {
+        StrNum = wcsrchr(ArgReal, L':');
+        StrNum++;
+    }
+
+    else if (nArgs == 3)
+    {
+        StrNum = ArgList[2];
+    }
+
+    Num = _wtoi(StrNum);
+
+    StringCchPrintfW(ValStr, ARR_MAX_SIZE, L"StateFlags%.4i", Num);
+
+    return ValStr;
+}
+
+BOOL DriverunProc(LPWSTR* ArgList, PWCHAR LogicalDrives)
+{
+    WCHAR DriveArgGathered[ARR_MAX_SIZE] = { 0 };
+    WCHAR TempText[ARR_MAX_SIZE] = { 0 };
+
+    StringCbCopyW(DriveArgGathered, sizeof(DriveArgGathered), ArgList[2]);
+
+    _wcsupr(DriveArgGathered);
+
+    if (wcslen(DriveArgGathered) == 1)
+    {
+        StringCbCatW(DriveArgGathered, sizeof(DriveArgGathered), L":");
+    }
+
+    WCHAR* SingleDrive = LogicalDrives;
+    WCHAR RealDrive[ARR_MAX_SIZE] = { 0 };
+    while (*SingleDrive)
+    {
+        if (GetDriveTypeW(SingleDrive) == ONLY_PHYSICAL_DRIVE)
+        {
+            StringCchCopyW(RealDrive, sizeof(RealDrive), SingleDrive);
+            RealDrive[wcslen(RealDrive) - 1] = '\0';
+            if (wcscmp(DriveArgGathered, RealDrive) == 0)
+            {
+                StringCbCopyW(wcv.DriveLetter, sizeof(wcv.DriveLetter), DriveArgGathered);
+                break;
+            }
+        }
+            SingleDrive += wcslen(SingleDrive) + 1;
+    }
+    if (wcslen(wcv.DriveLetter) == 0)
+    {
+        LoadStringW(GetModuleHandleW(NULL), IDS_ERROR_DRIVE, TempText, _countof(TempText));
+        MessageBoxW(NULL, TempText, L"Warning", MB_OK | MB_ICONERROR);
+        return FALSE;
+    }
+    
+    return TRUE;
+}
+
+void SagesetProc(int nArgs, PWCHAR ArgReal, LPWSTR* ArgList)
+{
+    WCHAR *ValStr = NULL;
+    INT_PTR DialogButtonSelect;
+        
+    ValStr = RealStageFlag(nArgs, ArgReal, ArgList);
+        
+    DialogButtonSelect = DialogBoxParamW(GetModuleHandleW(NULL), MAKEINTRESOURCEW(IDD_SAGERUN), NULL, SagesetDlgProc, 0);
+        
+    if (DialogButtonSelect == IDCANCEL)
+        return;
+
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Old ChkDsk Files", bv.ChkDskClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Temporary Files", bv.TempClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\RAPPS Files", bv.RappsClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Recycle Bin", bv.RecycleClean))
+        return;
+}
+
+void SagerunProc(int nArgs, PWCHAR ArgReal, LPWSTR* ArgList, PWCHAR LogicalDrives)
+{
+    WCHAR *ValStr = NULL;
+    WCHAR* SingleDrive = LogicalDrives;
+    INT_PTR DialogButtonSelect;
+        
+    ValStr = RealStageFlag(nArgs, ArgReal, ArgList);
```

Comment:
This function is a few lines above these functions using it.
Does everything in this binary need to use`RealStageFlag` (why Real, is there also an Imaginary StageFlag?)?
If not, then it can be removed from the shared header, and even made static.

Reply:
@learn-more `RealStageFlag` is actually a function which retrieves a number from the argument provided by the user and then gives a string with format like `StageFlagxxx`.

Reply:
yes, my point is that you drop all kinds of functions like that in a global / shared header, but they are only used in this file.